File: functions/src/notifications/eventUnreadPolicy.js
const { EVENT_EXPIRY_GRACE_DAYS } = require('../utils/constants');

const BRUSSELS = 'Europe/Brussels';

function asDate(value) {
  if (!value) return null;
  if (typeof value.toDate === 'function') return value.toDate();
  return value instanceof Date ? value : null;
}

// Calendar addition, rather than 24-hour addition, preserves the Brussels
// wall-clock time across daylight-saving transitions.
function eventUnreadUntil(dateFin) {
  const date = asDate(dateFin);
  if (!date) return null;
  const parts = new Intl.DateTimeFormat('en-CA', {
    timeZone: BRUSSELS, year: 'numeric', month: '2-digit', day: '2-digit',
    hour: '2-digit', minute: '2-digit', second: '2-digit', hourCycle: 'h23',
  }).formatToParts(date).reduce((out, part) => ({ ...out, [part.type]: part.value }), {});
  const localUtc = Date.UTC(Number(parts.year), Number(parts.month) - 1,
    Number(parts.day) + EVENT_EXPIRY_GRACE_DAYS, Number(parts.hour), Number(parts.minute), Number(parts.second));
  const offsetAt = (instant) => {
    const rendered = new Intl.DateTimeFormat('en-CA', {
      timeZone: BRUSSELS, year: 'numeric', month: '2-digit', day: '2-digit',
      hour: '2-digit', minute: '2-digit', second: '2-digit', hourCycle: 'h23',
    }).formatToParts(new Date(instant)).reduce((out, part) => ({ ...out, [part.type]: part.value }), {});
    return Date.UTC(rendered.year, rendered.month - 1, rendered.day,
      rendered.hour, rendered.minute, rendered.second) - instant;
  };
  let result = localUtc - offsetAt(localUtc);
  result = localUtc - offsetAt(result);
  return new Date(result + date.getMilliseconds());
}

function isUnreadEligibleEvent(operation = {}, now = new Date()) {
  const type = String(operation.type || '').trim().toLowerCase();
  const category = String(operation.event_category || operation.categorie || '').trim().toLowerCase();
  const status = String(operation.statut || '').trim().toLowerCase();
  if (type !== 'evenement' || category === 'piscine'
    || operation.deleted_at != null
    || !['ouvert', 'ferme', 'annule'].includes(status)) return false;
  const until = eventUnreadUntil(operation.date_fin);
  return !until || now.getTime() <= until.getTime();
}

module.exports = { asDate, eventUnreadUntil, isUnreadEligibleEvent, BRUSSELS };

File: lib/utils/event_unread_policy.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:timezone/data/latest.dart' as timezone_data;
import 'package:timezone/timezone.dart' as timezone;

DateTime? _operationEnd(Map<String, dynamic> operation) {
  final raw = operation['date_fin'];
  if (raw is Timestamp) return raw.toDate();
  if (raw is DateTime) return raw;
  return null;
}

/// Exact end of the seven Brussels-calendar-day window in which an event
/// discussion remains countable and must remain navigable from the event list.
DateTime? eventUnreadUntil(Map<String, dynamic> operation) {
  final end = _operationEnd(operation);
  if (end == null) return null;
  timezone_data.initializeTimeZones();
  final brussels = timezone.getLocation('Europe/Brussels');
  final localEnd = timezone.TZDateTime.from(end, brussels);
  return timezone.TZDateTime(
    brussels,
    localEnd.year,
    localEnd.month,
    localEnd.day + 7,
    localEnd.hour,
    localEnd.minute,
    localEnd.second,
    localEnd.millisecond,
    localEnd.microsecond,
  );
}

/// Canonical event-domain predicate shared by aggregate counts and rows.
///
/// Only actual CalyMob events in a member-visible lifecycle state can own an
/// event-discussion badge. Draft/deleted/other operation types and the legacy
/// piscine operation duplicate are excluded because they have no matching row
/// in the Events list.
bool isUnreadEligibleEvent(Map<String, dynamic> operation, DateTime now) {
  final type = operation['type']?.toString().trim().toLowerCase();
  final category = (operation['event_category'] ?? operation['categorie'])
      ?.toString()
      .trim()
      .toLowerCase();
  final status = operation['statut']?.toString().trim().toLowerCase();
  if (type != 'evenement' ||
      category == 'piscine' ||
      operation['deleted_at'] != null ||
      !const {'ouvert', 'ferme', 'annule'}.contains(status)) {
    return false;
  }
  final expiry = eventUnreadUntil(operation);
  return expiry == null ||
      !timezone.TZDateTime.from(
        now,
        timezone.getLocation('Europe/Brussels'),
      ).isAfter(expiry);
}

bool isCountableEventRegistration(Map<String, dynamic> registration) {
  final status =
      registration['registration_status']?.toString().trim().toLowerCase();
  return !const {'canceled', 'waitlisted', 'withdrawn'}.contains(status);
}
